README.md
# Nextcloud-Backup-Restore

Bash scripts for backup/restore of [Nextcloud](https://nextcloud.com/).

It is based on a Nextcloud installation using nginx and PostgreSQL/MariaDB (see the (German) tutorial [Nextcloud auf Ubuntu Server 22.04 LTS mit nginx, PostgreSQL/MariaDB, PHP, Let’s Encrypt, Redis und Fail2ban](https://decatec.de/home-server/nextcloud-auf-ubuntu-server-22-04-lts-mit-nginx-postgresql-mariadb-php-lets-encrypt-redis-und-fail2ban/)).\
The scripts can also be used when Apache is used as webserver.
This fork is based on the work of https://codeberg.org/DecaTec/Nextcloud-Backup-Restore and will be maintained on https://github.com/wagnbeu0/Nextcloud-Backup-Restore

## General information

For a complete backup of any Nextcloud instance, you'll have to backup these items:
- The Nextcloud **file directory** (usually */var/www/nextcloud*)
- The **data directory** of Nextcloud (it's recommended that this is *not* located in the web root, so e.g. */var/nextcloud_data*)
- The Nextcloud **database**
- Maybe a local external storage mounted into Nextcloud

With these scripts, all these elements can be included in a backup.

## Requirements

- *pigz* (https://zlib.net/pigz/) when using backup compression. If not installed already, it can be installed with `apt install pigz` (Debian/Ubuntu). If not available, you can use another compression algorithm (e.g. gzip)
- *rsync* when using incremental backups. If not installed already, it can be installed with `apt install rsync` (Debian/Ubuntu).

## Important notes about using the scripts

- After cloning or downloading the scripts, these need to be set up by running the script `setup.sh` (see below).
- If you do not want to use the automated setup, you can also use the file `NextcloudBackupRestore.conf.sample` as a starting point. Just make sure to rename the file when you are done (`cp NextcloudBackupRestore.conf.sample NextcloudBackupRestore.conf`)
- The configuration file `NextcloudBackupRestore.conf` has to be located in the same directory as the scripts for backup/restore.
- The scripts assume that Nextcloud's data directory is *not* a subdirectory of the Nextcloud installation (file directory). The general recommendation is that the data directory should not be located somewhere in the web folder of your webserver (usually */var/www/*), but in a different folder (e.g. */var/nextcloud_data*). For more information, see [here](https://docs.nextcloud.com/server/latest/admin_manual/installation/installation_wizard.html#data-directory-location-label).
- However, if your data directory *is* located under the Nextcloud file directory, you'll have to change the script configuration (file `NextcloudBackupRestore.conf` after running `setup.sh`) so that the data directory is not part of the backup/restore (otherwise, it would be copied twice).
- The scripts only backup the Nextcloud data directory and can backup a local external storage mounted into Nextcloud. If you have any other external storage mounted in Nextcloud (e.g. FTP), these files have to be handled separately.
- The scripts support nginx and Apache as webserver.
- The scripts support MariaDB/MySQL and PostgreSQL as database.
- You should have enabled 4 byte support (see [Nextcloud Administration Manual](https://docs.nextcloud.com/server/latest/admin_manual/configuration_database/mysql_4byte_support.html)) on your Nextcloud database. Otherwise, when you have *not* enabled 4 byte support, you have to edit the restore script, so that the database is not created with 4 byte support enabled (variable `dbNoMultibyte`).
- The scripts can exclude the Nextcloud data directory from backup and restore.\
**WARNING**: Excluding the data directory is **NOT RECOMMENDED** as it leaves the backup in an inconsistent state and may result in data loss!

## Setup

1. Clone the repository: `git clone https://github.com/wagnbeu0/Nextcloud-Backup-Restore`
2. Set permissions:
    - `chown -R root Nextcloud-Backup-Restore`
    - `cd Nextcloud-Backup-Restore`
    - `chmod 700 *.sh`
3. Call the (interactive) script for automated setup (this will create a file `NextcloudBackupRestore.conf` containing the desired configuration): `./setup.sh`
4. **Important**: Check this configuration file if everything was set up correctly (see *TODO* in the configuration file comments)
5. Start using the scripts: See sections *Backup* and *Restore* below

Keep in mind that the configuration file `NextcloudBackupRestore.conf` hast to be located in the same directory as the scripts for backup/restore, otherwise the configuration will not be found.

Some optional options are not configured using `setup.sh`, but are set to default values in `NextcloudBackupRestore.conf`. These are the "dangerous" options which usually should not be changed and are marked as 'OPTIONAL' in `NextcloudBackupRestore.conf`

## Backup

In order to create a backup, simply call the script *NextcloudBackup.sh* on your Nextcloud machine.
If this script is called without parameter, the backup is saved in a directory with the current time stamp in your main backup directory: As an example, this would be */media/hdd/nextcloud_backup/20170910_132703*.
The backup script can also be called with a parameter specifying the main backup directory, e.g. *./NextcloudBackup.sh /media/hdd/nextcloud_backup*. In this case, the directory specified will be used as main backup directory. 

You can also call this script by cron. Example (at 2am every night, with log output):

`0 2 * * * /path/to/scripts/Nextcloud-Backup-Restore/NextcloudBackup.sh  > /path/to/logs/Nextcloud-Backup-$(date +\%Y\%m\%d\%H\%M\%S).log 2>&1`

## Restore

Call *NextcloudRestore.sh* in order to restore a backup.\
When this script is called without parameters, it lists the backups available for restore.\
In order to restore a backup, call this script with a parameter specifying the name (i.e. timestamp) of the backup to be restored. In this example, this would be *20170910_132703*. The full command for a restore would be *./NextcloudRestore.sh 20170910_132703*.
You can also specify the main backup directory with a second parameter, e.g. *./NextcloudRestore.sh 20170910_132703 /media/hdd/nextcloud_backup*.


CHANGELOG.md
# Changelog

### 3.3.1
- fixed setup.sh: which has parameter `useCompression` for compressed backups statically set
- fixed setup.sh: which now asks if parameter `includeNextcloudDataDir` should be included in Backup
- fixed Nextcloud Backup.sh which now exludes data dir if not wanted in backup

## 3.3.0
- Use rsync as default for backups
- Disabled compressed backups to enable incremental sync backups
- Remove tar commands from code

## 3.2.0

### RSYNC
- Use RSync to be prepared for incremental backups

### General

- New configuration option `includeNextcloudDataDir` (default: `true`): This can be set to `false` in order to exclude Nextcloud's data directory from the backups (**not recommended**).
- Added hint to configuration file that `mariadb` and `mysql` are equivalent for the variable `databaseSystem`.
- Some configuration options in `NextcloudBackupRestore.conf` are optional, e.g. are set to default values by `setup.sh`. These are potentially dangerous options which should usually not be changed (e.g. `includeNextcloudDataDir`).

### Backup

- Use parameter `--default-character-set=utf8mb4` when dumping the database on MySQL/MariaDB and the 4-byte support is enabled in the `config.php`. 

### Restore

- When calling `NextcloudRestore.sh` without parameters, a list of available backups to restore is shown instead of error message.

## 3.0.3

### General
- Variable `includeUpdaterBackups` in `NextcloudBackupRestore.conf.sample`

### Backup
- Bugfix: The backup directory to use was not recognized when given by parameter

## 3.0.2

### Backup
- Bugfix: When calling the backup script by cron, the configuration file `NextcloudBackupRestore.conf` could not be found

## 3.0.1

### Backup
- Bugfix: Option to include/ignore backups from the Nextcloud updater was not set correctly during setup

## 3.0.0

### General
- Backup/restore scripts now use a central configuration file (`NextcloudBackupRestore.conf`)
- This configuration file includes all the settings which should be configured to match the specific Nextcloud instance. This simplifies the initial configuration of the backup/restore scripts
- The setup (`setup.sh`) creates this central configuration file (`NextcloudBackupRestore.conf`) rather than modifying the backup/restore scripts
- Bugfix: setup.sh won't set wrong database password if password contains slashes and/or backslashes

### Restore
- While restore, only the contents of Nextcloud's data directory gets removed, rather than the whole directory (useful when the data directory is stored directly in a mounted drive)

## 2.3.3

### Restore
- Update fingerprint after disabling maintenance mode

## 2.3.2

### Restore
- Bugfix: Compression command on restore

### Setup
- Hint for installing pigz when using compression
- Bugfix: Set variable `backupMainDir` correctly

## 2.3.1

### General
- Bugfix: Unbound variable when no parameters are supplied

## 2.3.0

### General
- The scripts now exit when any command fails.
- Defined the command for compression in the "TODO section" of the script for easier customization.
- Added section for setup in readme.
- Updated links. 
- Document requirement pigz when using compression.
- Formatting.

### Backup
- Bugfix: Fixed the double trailing slash for paths containing the variable `backupdir`.

## 2.2.0

### General
- Better handling of external data directory: Backup/restore of external data direcrory is done automatically if the variable `nextcloudLocalExternalDataDir` is set.

## 2.1.3

### General
- Added timestamps for every step.

## 2.1.2

### General
- Use pigz for compression.

## 2.1.1

### Backup
- Optimized cleaning up of old backups.

## 2.1.0

### General
- Added a variable *useCompression* to use compression (.tar.gz) for file and data directory (enabled by default, this was also the default behavior before this option was implemented). 
- You can disable compression of these directories (.tar) if you use some other (archiving) backup mechanism which utilizes the Nextcloud backup and restore scripts (e.g. Borg Backup, see here: https://codeberg.org/DecaTec/Linux-Server-Backup)

## 2.0.0

### General
- Added script (`setup.sh`) for an (more or less) automated setup of the backup and restore scripts (utilizing OCC in order to gather information about Nextcloud instance).

## 1.1.1

### Backup
- When a backup is cancelled, the webserver is restartet automatically.

## 1.1.0

### Backup
- New variable *ignoreUpdaterBackups*: When set to true, the backups of Nextcloud's updater are not included in the backups (default: *false*).

## 1.0.0

### General
- Versioning of Nextcloud-Backup-Restore.
- The database system (MySQL/MariaDB or PostgreSQL) is configured in the variable area of the scripts, so it's not necessary to comment/uncomment the specific database commands.
- Special characters for the database password can be used now.
- Single quotes for variables.

### Restore
- The commands for restoring the database are checked at the beginning of the script. Is the specific database system is not installed, the restore is cancelled.
- The default main backup directory now is the same as in the backup script.


NextcloudBackup.sh
#!/bin/bash

#
# Bash script for creating backups of Nextcloud. This Script is based on the work of https://codeberg.org/DecaTec/Nextcloud-Backup-Restore
#
# Version 3.3.1
#
# Requirements:
#	- pigz (https://zlib.net/pigz/) for using backup compression. If not available, you can use another compression algorithm (e.g. gzip)
#	- rsync for using noncopmpressed backups.
#
# Supported database systems:
# 	- MySQL/MariaDB
# 	- PostgreSQL
#
# Usage:
# 	- With backup directory specified in the script:  ./NextcloudBackup.sh
# 	- With backup directory specified by parameter: ./NextcloudBackup.sh <backupDirectory> (e.g. ./NextcloudBackup.sh /media/hdd/nextcloud_backup)
#
# The script is based on an installation of Nextcloud using nginx and MariaDB, see https://decatec.de/home-server/nextcloud-auf-ubuntu-server-22-04-lts-mit-nginx-postgresql-mariadb-php-lets-encrypt-redis-und-fail2ban/
#


# Make sure the script exits when any command fails
set -Eeuo pipefail

# Variables
working_dir=$(cd $(dirname "${BASH_SOURCE[0]}") && pwd)
configFile="${working_dir}/NextcloudBackupRestore.conf"   # Holds the configuration for NextcloudBackup.sh and NextcloudRestore.sh
_backupMainDir=${1:-}

# Function for error messages
errorecho() { cat <<< "$@" 1>&2; }

#
# Check for root
#
if [ "$(id -u)" != "0" ]
then
	errorecho "ERROR: This script has to be run as root!"
	exit 1
fi

#
# Check if config file exists
#
if [ ! -f "${configFile}" ]
then
	errorecho "ERROR: Configuration file $configFile cannot be found!"
	errorecho "Please make sure that a configuratrion file '$configFile' is present in the main directory of the scripts."
	errorecho "This file can be created automatically using the setup.sh script."
	exit 1
fi

source "$configFile" || exit 1  # Read configuration variables

if [ -n "$_backupMainDir" ]; then
	backupMainDir=$(echo $_backupMainDir | sed 's:/*$::')
fi

currentDate=$(date +"%Y%m%d_%H%M%S")

# The actual directory of the current backup - this is a subdirectory of the main directory above with a timestamp
backupDir="${backupMainDir}/${currentDate}"

function DisableMaintenanceMode() {
	echo "$(date +"%H:%M:%S"): Switching off maintenance mode..."
	sudo -u "${webserverUser}" php ${nextcloudFileDir}/occ maintenance:mode --off
	echo "Done"
	echo
}

function occ_get() {
	sudo -u "${webserverUser}" php ${nextcloudFileDir}/occ config:system:get "$1"
}

if [ "${databaseSystem,,}" = "mysql" ] || [ "${databaseSystem,,}" = "mariadb" ]; then
  fourByteSupport=$(occ_get mysql.utf8mb4) || (errorecho "ERROR: OCC config:system:get call failed!" && exit 1)
fi

# Capture CTRL+C
trap CtrlC INT

function CtrlC() {
	read -p "Backup cancelled. Keep maintenance mode? [y/n] " -n 1 -r
	echo

	if ! [[ $REPLY =~ ^[Yy]$ ]]
	then
		DisableMaintenanceMode
	else
    echo "Maintenance mode still enabled."
	fi

	echo "Starting web server..."
	systemctl start "${webserverServiceName}"
	echo "Done"
	echo

	exit 1
}

#
# Print information
#
echo "Backup directory: ${backupMainDir}"

#
# Check if backup dir already exists
#
if [ ! -d "${backupDir}" ]
then
	mkdir -p "${backupDir}"
else
	errorecho "ERROR: The backup directory ${backupDir} already exists!"
	exit 1
fi

#
# Set maintenance mode
#
echo "$(date +"%H:%M:%S"): Set maintenance mode for Nextcloud..."
sudo -u "${webserverUser}" php ${nextcloudFileDir}/occ maintenance:mode --on
echo "Done"
echo

#
# Stop web server
#
echo "$(date +"%H:%M:%S"): Stopping web server..."
systemctl stop "${webserverServiceName}"
echo "Done"
echo

#
# Backup file directory
#
echo "$(date +"%H:%M:%S"): Creating backup of Nextcloud file directory..."

if [ "$useCompression" = true ] ; then
	if [ "$includeNextcloudDataDir" = false ]; then
		echo "$(date +"%H:%M:%S"): Creating compressed tar backup of Nextcloud file directory without data subfolder..."
		`$compressionCommand "${backupDir}/${fileNameBackupFileDir}" --exclude="./data/*" -C "${nextcloudFileDir}" .`
	else
		echo "$(date +"%H:%M:%S"): Creating compressed tar backup of Nextcloud file directory with data subfolder..."
		`$compressionCommand "${backupDir}/${fileNameBackupFileDir}" -C "${nextcloudFileDir}" .`
	fi
else
	if [ "$includeNextcloudDataDir" = false ]; then
		if [ -d "${backupMainDir}/latest" ]; then
			echo "$(date +"%H:%M:%S"): Create incremental backup using rsync without data subfolder..."
			rsync -aug "${nextcloudFileDir}/" --link-dest "${backupMainDir}/latest/" "${backupDir}/${folderNameBackupFileDir}/" --exclude "data/*" 
		else
			echo "$(date +"%H:%M:%S"): Create full backup using rsync without data subfolder..."
			rsync -aug "${nextcloudFileDir}/" "${backupDir}/${folderNameBackupFileDir}/" --exclude "data/*" 
		fi
	else
		if [ -d "${backupMainDir}/latest" ]; then
			echo "$(date +"%H:%M:%S"): Create incremental backup using rsync with data subfolder..."
			rsync -aug "${nextcloudFileDir}/" --link-dest "${backupMainDir}/latest/" "${backupDir}/${folderNameBackupFileDir}/" 
		else
			echo "$(date +"%H:%M:%S"): Create full backup using rsync with data subfolder..."
			rsync -aug "${nextcloudFileDir}/" "${backupDir}/${folderNameBackupFileDir}/" 
		fi
	fi
fi

echo "Done"
echo

#
# Backup data directory
#
if [ "$includeNextcloudDataDir" = false ]; then
	echo "$(date +"%H:%M:%S"): Ignoring backup of Nextcloud data directory!"
elif [[ "${nextcloudDataDir}" = "${nextcloudFileDir}"* ]] && [ "$includeNextcloudDataDir" = true ]; then
	echo "$(date +"%H:%M:%S"): Skipping backup of Nextcloud data directory (already included in file directory backup)!"
else
	echo "$(date +"%H:%M:%S"): Creating backup of Nextcloud data directory..."

	if [ "$includeUpdaterBackups" = false ] ; then
		echo "Ignoring Nextcloud updater backup directory"

		if [ "$useCompression" = true ] ; then
			echo "$(date +"%H:%M:%S"): Creating compressed tar backup"
			`$compressionCommand "${backupDir}/${fileNameBackupDataDir}"  --exclude="updater-*/backups/*" -C "${nextcloudDataDir}" .`
		else
			if [ -d "${backupMainDir}/latest" ]; then
			echo "$(date +"%H:%M:%S"): Creating incremental rsync backup"
			rsync -aug "${nextcloudDataDir}/" --link-dest "${backupMainDir}/latest/" "${backupDir}/${folderNameBackupDataDir}/" --exclude "updater-*/backups/*"			
			else
			rsync -aug "${nextcloudDataDir}/" "${backupDir}/${folderNameBackupDataDir}/" --exclude "updater-*/backups/*"
			fi	
		fi
	else
		if [ "$useCompression" = true ] ; then
			echo "$(date +"%H:%M:%S"): Creating compressed tar backup"
			`$compressionCommand "${backupDir}/${fileNameBackupDataDir}"  -C "${nextcloudDataDir}" .`
		else
			if [ -d "${backupMainDir}/latest" ]; then
			echo "$(date +"%H:%M:%S"): Creating incremental rsync backup"
			rsync -aug "${nextcloudDataDir}/" --link-dest "${backupMainDir}/latest/" "${backupDir}/${folderNameBackupDataDir}/"			
			else
			echo "$(date +"%H:%M:%S"): Creating incremental rsync backup"
			rsync -aug "${nextcloudDataDir}/" "${backupDir}/${folderNameBackupDataDir}/"
			fi
		fi
	fi
fi

echo "Done"
echo

#
# Backup local external storage.
#
if [ ! -z "${nextcloudLocalExternalDataDir+x}" ] ; then
	echo "$(date +"%H:%M:%S"): Creating backup of Nextcloud local external storage directory..."

	if [ "$useCompression" = true ] ; then
		`$compressionCommand "${backupDir}/${fileNameBackupExternalDataDir}"  -C "${nextcloudLocalExternalDataDir}" .`
	else
		tar -cpf "${backupDir}/${fileNameBackupExternalDataDir}"  -C "${nextcloudLocalExternalDataDir}" .
	fi

	echo "Done"
	echo
fi

#
# Backup DB
#
if [ "${databaseSystem,,}" = "mysql" ] || [ "${databaseSystem,,}" = "mariadb" ]; then
  	echo "$(date +"%H:%M:%S"): Backup Nextcloud database (MySQL/MariaDB)..."

	if ! [ -x "$(command -v mysqldump)" ]; then
		errorecho "ERROR: MySQL/MariaDB not installed (command mysqldump not found)."
		errorecho "ERROR: No backup of database possible!"
	else
		if [ $fourByteSupport = "true" ]; then
			mysqldump --single-transaction --default-character-set=utf8mb4 -h localhost -u "${dbUser}" -p"${dbPassword}" "${nextcloudDatabase}" > "${backupDir}/${fileNameBackupDb}"
		else
			mysqldump --single-transaction -h localhost -u "${dbUser}" -p"${dbPassword}" "${nextcloudDatabase}" > "${backupDir}/${fileNameBackupDb}"
		fi
	fi

	echo "Done"
	echo
elif [ "${databaseSystem,,}" = "postgresql" ] || [ "${databaseSystem,,}" = "pgsql" ]; then
	echo "$(date +"%H:%M:%S"): Backup Nextcloud database (PostgreSQL)..."

	if ! [ -x "$(command -v pg_dump)" ]; then
		errorecho "ERROR: PostgreSQL not installed (command pg_dump not found)."
		errorecho "ERROR: No backup of database possible!"
	else
		PGPASSWORD="${dbPassword}" pg_dump "${nextcloudDatabase}" -h localhost -U "${dbUser}" -f "${backupDir}/${fileNameBackupDb}"
	fi

	echo "Done"
	echo
fi

#
# Start web server
#
echo "$(date +"%H:%M:%S"): Starting web server..."
systemctl start "${webserverServiceName}"
echo "Done"
echo

#
# Disable maintenance mode
#
DisableMaintenanceMode

#
# Delete old backups
#
if [ ${maxNrOfBackups} != 0 ]
then
	nrOfBackups=$(ls -l ${backupMainDir} | grep -c ^d)

	if [ ${nrOfBackups} -gt ${maxNrOfBackups} ]
	then
		echo "$(date +"%H:%M:%S"): Removing old backups..."
		ls -t ${backupMainDir} | tail -$(( nrOfBackups - maxNrOfBackups )) | while read -r dirToRemove; do
			echo "${dirToRemove}"
			rm -r "${backupMainDir}/${dirToRemove:?}"
			echo "Done"
			echo
		done
	fi
fi

echo
echo "DONE!"
echo "$(date +"%H:%M:%S"): Backup created: ${backupDir}"

# Set this backup as latest
rm -rf "${backupMainDir}/latest"
ln -s "${backupDir}" "${backupMainDir}/latest"

set +Eeuo pipefail


setup.sh
#!/bin/bash

#
# Bash script for an easy setup of Nextcloud backup/restore scripts.
#
# Version 3.3.1
#
# Usage:
#   - call the setup.sh script
#   - Enter the required information
#   - A central configuration file `NextcloudBackupRestore.conf` will be created to match you Nextcloud instance.
#   - This configuration file then is used by the backup/restore scripts.
#
# The script is based on an installation of Nextcloud using nginx and MariaDB, see https://decatec.de/home-server/nextcloud-auf-ubuntu-server-22-04-lts-mit-nginx-postgresql-mariadb-php-lets-encrypt-redis-und-fail2ban/
#

#
# IMPORTANT
# The setup.sh script automated the configuration for the backup/restore scripts (file `NextcloudBackupRestore.conf`).
# However, you should always check this configuration BEFORE executing these!
#

# Make sure the script exits when any command fails
set -Eeuo pipefail

#
# Pre defined variables
#
backupMainDir='/backup'
nextcloudFileDir='/var/www/html/nextcloud'
webserverUser='www-data'
webserverServiceName='nginx'
useCompression=false
includeUpdaterBackups=false
includeNextcloudDataDir=true

NextcloudBackupRestoreConf='NextcloudBackupRestore.conf'  # Holds the configuration for NextcloudBackup.sh and NextcloudRestore.sh

# Function for error messages
errorecho() { cat <<< "$@" 1>&2; }

#
# Check for root
#
if [ "$(id -u)" != "0" ]
then
	errorecho "ERROR: This script has to be run as root!"
	exit 1
fi

#
# Gather information
#
clear

echo "Enter the directory to which the backups should be saved."
echo "Default: ${backupMainDir}"
echo ""
read -p "Enter a directory or press ENTER if the backup directory should be ${backupMainDir}: " BACKUPMAINDIR

[ -z "$BACKUPMAINDIR" ] ||  backupMainDir=$BACKUPMAINDIR
clear

echo "Enter the path to the Nextcloud file directory."
echo "Usually: ${nextcloudFileDir}"
echo ""
read -p "Enter a directory or press ENTER if the file directory is ${nextcloudFileDir}: " NEXTCLOUDFILEDIRECTORY

[ -z "$NEXTCLOUDFILEDIRECTORY" ] ||  nextcloudFileDir=$NEXTCLOUDFILEDIRECTORY
clear

echo "Enter the webserver user."
echo "Usually: ${webserverUser}"
echo ""
read -p "Enter an new user or press ENTER if the webserver user is ${webserverUser}: " WEBSERVERUSER

[ -z "$WEBSERVERUSER" ] ||  webserverUser=$WEBSERVERUSER
clear

echo "Enter the webserver service name."
echo "Usually: nginx or apache2"
echo ""
read -p "Enter an new webserver service name or press ENTER if the webserver service name is ${webserverServiceName}: " WEBSERVERSERVICENAME

[ -z "$WEBSERVERSERVICENAME" ] ||  webserverServiceName=$WEBSERVERSERVICENAME
clear

echo ""
read -p "Should the backed up data be compressed (pigz should be installed in the machine)? [N/y]: " USECOMPRESSION
if [ "$USECOMPRESSION" == 'Y' ] ; then
  useCompression=true
fi

clear

echo ""
read -p "Should the backup include the data folder? RECOMMENDED to avoid inconsistent backups [Y/n]: " INCLUDENEXTCLOUDDATADIR
if [ "$INCLUDENEXTCLOUDDATADIR" == 'n' ] ; then
  includeNextcloudDataDir=false
fi

clear

echo ""
read -p "Should the backups created by the Nextcloud updater be included in the backups (usually not necessary)? [y/N]: " INCLUDEPDATERBACKUPS

if [ "$INCLUDEPDATERBACKUPS" == 'y' ] ; then
  includeUpdaterBackups=true
fi

clear

echo "How many backups should be kept?"
echo "If this is set to '0', no backups will be deleted."
echo ""
read -p "How many backups should be kept (default: '0'): " MAXNUMBEROFBACKUPS

maxNrOfBackups=0

if ! [ -z "$MAXNUMBEROFBACKUPS" ]  && ! [[ "$MAXNUMBEROFBACKUPS" =~ ^[0-9]+$ ]] ; then
  echo "ERROR: Number of backups must be a positive integer!"
  echo ""
  echo "ABORTING!"
  echo "No file has been altered."
  exit 1
fi

[ -z "$MAXNUMBEROFBACKUPS" ] ||  maxNrOfBackups=$MAXNUMBEROFBACKUPS
clear

echo "Backup directory: ${backupMainDir}"
echo "Nextcloud file directory: ${nextcloudFileDir}"
echo "Webserver user: ${webserverUser}"
echo "Webserver service name: ${webserverServiceName}"

if [ "$useCompression" = true ] ; then
	echo "Compression: yes"
else
  echo "Compression: no"
fi

if [ "$includeNextcloudDataDir" = true ] ; then
	echo "Include Nextcloud Data dir: yes"
else
  echo "Include Nextcloud Data dir: no"
fi

if [ "$includeUpdaterBackups" = true ] ; then
	echo "Include backups from the Nextcloud updater: yes"
else
  echo "Include backups from the Nextcloud updater: no"
fi

echo "Number of backups to keep (0: keep all backups): ${maxNrOfBackups}"
echo ""

read -p "Is the information correct? [y/N] " CORRECTINFO

if [ "$CORRECTINFO" != 'y' ] ; then
  echo ""
  echo "ABORTING!"
  echo "No file has been altered."
  exit 1
fi

function occ_get() {
	sudo -u "${webserverUser}" php ${nextcloudFileDir}/occ config:system:get "$1"
}

# Make test call to OCC
occ_get datadirectory >/dev/null 2>&1

if [ $? -ne 0 ]; then
  echo ""
  echo "Error calling OCC: Please check if the information provided was correct."
  echo "ABORTING!"
  echo "No file has been altered."
  exit 1
fi

#
# Read data from OCC and write to config file.
#

if [ -e "$NextcloudBackupRestoreConf" ] ; then
  echo -e "\n\nSaving existing $NextcloudBackupRestoreConf to ${NextcloudBackupRestoreConf}_bak"
  cp --force "$NextcloudBackupRestoreConf" "${NextcloudBackupRestoreConf}_bak"
fi

echo ""
echo ""
echo "Creating $NextcloudBackupRestoreConf to match your Nextcloud instance..."
echo ""

# Nextcloud data dir
nextcloudDataDir=$(occ_get datadirectory)

# Database system
databaseSystem=$(occ_get dbtype)

# PostgreSQL is identified as pgsql
if [ "${databaseSystem,,}" = "pgsql" ]; then
  databaseSystem='postgresql';
fi

# Database
nextcloudDatabase=$(occ_get dbname)

# Database user
dbUser=$(occ_get dbuser)

# Database password
dbPassword=$(occ_get dbpassword)

# Folder names for backup files
folderNameBackupFileDir='nextcloud-filedir'
folderNameBackupDataDir='nextcloud-datadir'

# File names for backup files
fileNameBackupFileDir='nextcloud-filedir.tar'
fileNameBackupDataDir='nextcloud-datadir.tar'

if [ "$useCompression" = true ] ; then
	fileNameBackupFileDir='nextcloud-filedir.tar.gz'
	fileNameBackupDataDir='nextcloud-datadir.tar.gz'
fi

fileNameBackupExternalDataDir=''

if [ ! -z "${nextcloudLocalExternalDataDir+x}" ] ; then
	fileNameBackupExternalDataDir='nextcloud-external-datadir.tar'

	if [ "$useCompression" = true ] ; then
		fileNameBackupExternalDataDir='nextcloud-external-datadir.tar.gz'
	fi
fi

fileNameBackupDb='nextcloud-db.sql'

{ echo "# Configuration for Nextcloud-Backup-Restore scripts"
  echo ""
  echo "# TODO: The main backup directory"
  echo "backupMainDir='$backupMainDir'"
  echo ""
  echo "# TODO: Use compression for file/data dir"
  echo "# When this is the only script for backups, it is recommend to enable compression."
  echo "# If the output of this script is used in another (compressing) backup (e.g. borg backup),"
  echo "# you should probably disable compression here and only enable compression of your main backup script."
  echo "useCompression=$useCompression"
  echo ""
  echo "# TOOD: The bare tar command for using compression while backup."
  echo "# Use 'tar -cpzf' if you want to use gzip compression."
  echo "compressionCommand='tar -I pigz -cpf'"
  echo ""
  echo "# TOOD: The bare tar command for using compression while restoring."
  echo "# Use 'tar -xmpzf' if you want to use gzip compression."
  echo "extractCommand='tar -I pigz -xmpf'"
  echo ""
  echo "# TODO: File names for backup files"
  echo "fileNameBackupFileDir='$fileNameBackupFileDir'"
  echo "fileNameBackupDataDir='$fileNameBackupDataDir'"
  echo "fileNameBackupExternalDataDir='$fileNameBackupExternalDataDir'"
  echo "fileNameBackupDb='$fileNameBackupDb'"
  echo "folderNameBackupFileDir='$folderNameBackupFileDir'"
  echo "folderNameBackupDataDir='$folderNameBackupDataDir'"
  echo ""
  echo "# TODO: The directory of your Nextcloud installation (this is a directory under your web root)"
  echo "nextcloudFileDir='$nextcloudFileDir'"
  echo ""
  echo "# The directory of your Nextcloud data directory (outside the Nextcloud file directory)"
  echo "# If your data directory is located under Nextcloud's file directory (somewhere in the web root),"
  echo "# the data directory will not be a separate part of the backup but included in the file directory backup."
  echo "nextcloudDataDir='$nextcloudDataDir'"
  echo ""
  echo "# TODO: The directory of your Nextcloud's local external storage."
  echo "# Uncomment if you use local external storage."
  echo "#nextcloudLocalExternalDataDir='/var/nextcloud_external_data'"
  echo ""
  echo "# TODO: The service name of the web server. Used to start/stop web server (e.g. 'systemctl start <webserverServiceName>')"
  echo "webserverServiceName='$webserverServiceName'"
  echo ""
  echo "# TODO: Your web server user"
  echo "webserverUser='$webserverUser'"
  echo ""
  echo "# TODO: The name of the database system (one of: mysql, mariadb, postgresql)"
  echo "# 'mysql' and 'mariadb' are equivalent, so when using 'mariadb', you could also set this variable to 'mysql' and vice versa."
  echo "databaseSystem='$databaseSystem'"
  echo ""
  echo "# TODO: Your Nextcloud database name"
  echo "nextcloudDatabase='$nextcloudDatabase'"
  echo ""
  echo "# TODO: Your Nextcloud database user"
  echo "dbUser='$dbUser'"
  echo ""
  echo "# TODO: The password of the Nextcloud database user"
  echo "dbPassword='$dbPassword'"
  echo ""
  echo "# TODO: The maximum number of backups to keep (when set to 0, all backups are kept)"
  echo "maxNrOfBackups=$maxNrOfBackups"
  echo ""
  echo "# TODO: Setting to include/exclude the backup directory of the Nextcloud updater"
  echo "# Set to true in order to include the backups of the Nextcloud updater"
  echo "includeUpdaterBackups=$includeUpdaterBackups"
  echo ""
  echo "# OPTIONAL: Setting to include/exclude the Nextcloud data directory"
  echo "# Set to false to exclude the Nextcloud data directory from backup"
  echo "# WARNING: Excluding the data directory is NOT RECOMMENDED as it leaves the backup in an inconsistent state and may result in data loss!"
  echo "includeNextcloudDataDir=$includeNextcloudDataDir"

} > ./"${NextcloudBackupRestoreConf}"

echo ""
echo "Done!"
echo ""
echo ""
echo "IMPORTANT: Please check $NextcloudBackupRestoreConf if all variables were set correctly BEFORE running the backup/restore scripts!"

if [ "$useCompression" = true ] ; then
  echo ""
	echo "As compression should be used for backups, please make sure that 'pigz' is installed (e.g. for Debian/Ubuntu: apt install pigz)"
fi

echo ""
echo ""

set +Eeuo pipefail
